ufs-community/ufs-weather-model
# CCPP tendencies bugfixes, global restart reproducibility, halo boundary update in dycore

## Description

This PR
- updates the submodule pointers for fv3atm, GFDL_atmos_cubed_sphere and ccpp-physics for the following changes:
    - bugfixes for CCPP tendencies output (from @grantfirl)
    - global restart reproducibility (from @SMoorthi-emc)
    - bugfix for halo boundary layer update in FV3 dycore (from @chan-hoo)
- moves directory `parm` into directory `tests`
- updates regression test run scripts for HAFS and SAR to copy input data (`diag_table`, `field_table`, initial conditions) from the regression baseline instead of a temporary directory on hera
- note: the following existing diag_table files in the regression test baseline directory `FV3_input_data_gsd` have been updated to match the requirements for running cycled experiments with the GSL physics (@jkhender @joeolson42 FYI)
```
diag_table_gf_thompson
diag_table_gsd_noah
diag_table_gsd_ruc
diag_table_gsd_ruc_diag3d
```

### Issue(s) addressed

- (partly) https://github.com/ufs-community/ufs-weather-model/issues/196
- https://github.com/NOAA-EMC/fv3atm/issues/159

## Testing

The above changes require new baselines for all regional tests and for the IAU/read increment tests (due to the halo update bugfix in the fv3 dycore), and for all tests using the GSL physics (due to the updates to the `diag_table` files). I first created new baselines on hera with both Intel and GNU and compared them against the existing baselines to make sure that only tests/files changed that are expected to change. I then copied the new baselines to the regression test baseline directory with date tag `20200929`.

Note that this `20200929` baseline contains several new/updated files that need to be copied to other machines before creating new baselines there:
```
FV3_input_data_gsd/
    diag_table_gf_thompson
    diag_table_gsd_noah
    diag_table_gsd_ruc
    diag_table_gsd_ruc_diag3d
FV3_input_data_hafs/*
FV3_input_data_sar/*
FV3_input_data_sar_25km/*
```
Tests that have different results compared to the 20200923 baseline because of the halo update bugfix in the dycore, the updates to the GSL diag tables, Moorthi's reproducibility bugfix and the CCPP diagnostic tendency bugfix:
```
# test results change due to the updates to the diag tables for GSL physics (phyf*.nc only, no changes to dynf*.nc)
fv3_gsd_ccpp
fv3_gsd_debug_ccpp
fv3_hrrr_ccpp
fv3_rap_ccpp
fv3_rrfs_v1beta_ccpp
fv3_rrfs_v1beta_debug_ccpp
fv3_thompson_ccpp
fv3_thompson_debug_ccpp
fv3_thompson_no_aero_ccpp
fv3_thompson_no_aero_debug_ccpp

# test results change due to the updates to the diag tables for GSL physics and the CCPP diagnostic tendency bugfix
fv3_gsd_diag3d_debug_ccpp

# test results change due to the halo boundary update bugfix
fv3_regional_c768_ccpp
fv3_regional_control_ccpp
fv3_regional_quilt_ccpp
fv3_regional_restart_ccpp

# test results change due to Moorthi's global restart reproducibility bugfix - note that I have NOT done a manual test if this really, really fixes the global restart reproducibility issues, nor did I fix the global restart test
fv3_restart_ccpp

# test results change either due to Moorthi's global restart reproducibility bugfix or due to the halo boundary update bugfix (not sure)
fv3_iau_ccpp
fv3_read_inc_ccpp
```
That is, 18 tests in total - the remaining 60-18 = 42 tests are unchanged.

### Final regression testing against new baseline (20200929):

Regression tests passed on
- hera.intel
- hera.gnu
- wcoss_dell_p3
- wcoss_cray
- orion.intel
Regression test logs updated in the PR.

I am in the process of creating new regression test baselines and verifying against them on cheyenne.intel and cheyenne.gnu, but we do not have to wait for this process to complete (cheyenne is tier-2).

### Manual testing

- code was compiled on gaea.intel to make sure that the changes to `compile.sh` work there as well
- code was compiled on jet.intel to make sure that the changes to `compile.sh` work there as well

## Dependencies

https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere/pull/40 (still being worked on)
https://github.com/NCAR/ccpp-physics/pull/489
https://github.com/NOAA-EMC/fv3atm/pull/178
https://github.com/ufs-community/ufs-weather-model/pull/208

Corresponding ufs-s2s-model PR: https://github.com/ufs-community/ufs-s2s-model/pull/200
